### PR TITLE
HPCC-18095 Refactor VarOffset segmonitors so they can be serialized

### PR DESCRIPTION
**Pull request by richardkchapman**
@ghalliday For initial comment - not ready for full review or merge yet

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Test case being cleaned up for addition to suite.

### CODE REVIEW

File: rtl/eclrtl/rtlkey.cpp [hunk 1]
```diff
@@ -815,56 +826,88 @@ class CIndirectKeySegmentMonitor : implements IKeySegmentMonitor, public CInterf
...
     {
-        base->copy((byte *)expandedRow + offset, offsetTranslator->getSegmentBase(rawRight));
+        throwUnexpected();
     }
     virtual int docompareraw(const void * left, const void * right) const
     {
-        return base->docompare(offsetTranslator->getSegmentBase(left), offsetTranslator->getSegmentBase(right));
+        throwUnexpected();
     }
+
```
Comment (ghalliday) [inline rtl/eclrtl/rtlkey.cpp]: Should you also fail on various other functions - increment, setLow, setHigh, or implement them?

Comment (richardkchapman) [inline rtl/eclrtl/rtlkey.cpp]: They are all irrelevant until support index creation I think.

Comment (richardkchapman) [inline rtl/eclrtl/rtlkey.cpp]: Technically I probably should fail on them (I can't implement them without being able to translate offsets on buffer side). But code I am replacing did not...

Could open a separate Jira

File: rtl/eclrtl/rtlkey.cpp [hunk 2]
```diff
@@ -815,56 +826,88 @@ class CIndirectKeySegmentMonitor : implements IKeySegmentMonitor, public CInterf
...
+        mb.read(fieldIdx);
+    }
+
+    virtual bool matches(const RtlRow *keyval) const
+    {
+        return base->matchesBuffer(getSegmentBase(keyval));
     }
 
-    virtual bool matches(const void *keyval) const
+    virtual unsigned queryFieldNo() const { return fieldIdx+1; }
```
Comment (ghalliday) [inline rtl/eclrtl/rtlkey.cpp]: Shouldn't this be
    virtual unsigned numFieldsRequired() const
an override attribute would help.

Comment (richardkchapman) [inline rtl/eclrtl/rtlkey.cpp]: Hmmm - probably. Which begs the question of how tests passed

File: rtl/eclrtl/rtlrecord.cpp [hunk 3]
```diff
@@ -280,17 +280,23 @@ RtlRecord::~RtlRecord()
     delete nameMap;
 }
 
-
-void RtlRecord::calcRowOffsets(size_t * variableOffsets, const void * _row) const
+void RtlRecord::calcRowOffsets(size_t * variableOffsets, const void * _row, unsigned numFieldsUsed) const
 {
     const byte * row = static_cast<const byte *>(_row);
-    for (unsigned i = 0; i < numVarFields; i++)
+    unsigned maxVarField = (numFieldsUsed>=numFields) ? numVarFields : whichVariableOffset[numFields];
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.cpp]: whichVariableOffset[numFieldsUsed];

File: system/jhtree/CMakeLists.txt [hunk 4]
```diff
@@ -28,7 +28,7 @@ project( jhtree )
 set (    SRCS 
          hlzw.cpp 
          ctfile.cpp 
-         jhrtl.cpp 
+#         jhrtl.cpp
```
Comment (ghalliday) [inline system/jhtree/CMakeLists.txt]: also delete the file (and remove this line)?

File: thorlcr/activities/diskread/thdiskreadslave.cpp [hunk 5]
```diff
@@ -51,13 +52,22 @@ class CDiskReadSlaveActivityRecord : public CDiskReadSlaveActivityBase, implemen
     bool grouped;
     bool isFixedDiskWidth;
     size32_t diskRowMinSz;
+    const RtlRecord *recInfo = nullptr;
+    unsigned numOffsets = 0;
```
Comment (ghalliday) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: Should this default to 1?

Comment (richardkchapman) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: No - if all fields are fixed there's no need to do any offset calculation

Comment (richardkchapman) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: Cancel that, I misread.
Don't think it really matters. It's not meaningful until recInfo set up

**Pull request merged**